[수달] 1, 2단계 영화 극장 선택 제출합니다.

otter66 opened the pull request:
안녕하세요! 페로로님!
모바일 수달입니다!
이번 미션동안 잘 부탁드리겠습니다!

rhthrhrl0 committed:
merge: 좌석 예매 앱과 영화 예매 앱 병합

rhthrhrl0 committed:
docs: 기능 목록 문서 작성

rhthrhrl0 committed:
feat(MovieListFragment): 영화 목록 프래그먼트 추가

rhthrhrl0 committed:
feat(ReservationListFragment): 영화 예약 목록 프래그먼트 추가

rhthrhrl0 committed:
feat(ManinActivity): 바텀 네비게이션 기능 추가

rhthrhrl0 committed:
feat(SettingFragment): 설정 화면 추가

rhthrhrl0 committed:
feat(ReservationConfirmActivity): 예약내역을 레포지토리에 추가하는 기능 구현

rhthrhrl0 committed:
feat(reservation): 예약 목록 아이템 클릭 시 화면 전환 기능 추가

rhthrhrl0 committed:
refactor: 클래스명 변경

- reservation과 ticket이 혼용되는 것을 방지하기 위해 변경

rhthrhrl0 committed:
feat(setting): 푸시 알람 수신 기능 화면 추가

rhthrhrl0 committed:
feat: 권한 요청 기능 추가

rhthrhrl0 committed:
feat: 영화 상영 시간 30분 전 알림 기능 추가

rhthrhrl0 committed:
feat: 알람 수신 설정 기능 추가

rhthrhrl0 committed:
feat: 알림 기능 추가

rhthrhrl0 committed:
fix: 알림 수신 여부 값이 반영되지 않는 오류 해결

re4rk committed:
fix: 알림 권한이 없을 때 설정 창에서 푸시 알림 수신이 선택되지 않도록 수정

rhthrhrl0 committed:
refactor(MainActivity): 메인 액티비티 onCreate 함수 분리

rhthrhrl0 committed:
docs: 기능 목록 문서 작성

rhthrhrl0 committed:
fix: 예매 목록 업데이트 안되던 버그 수정

laco-dev commented on app/src/main/AndroidManifest.xml:
```diff
@@ -11,6 +17,10 @@
         tools:targetApi="31">
+        <receiver
+            android:name=".ui.confirm.AlarmReceiver"
+            android:process=":remote"></receiver>
```
process를 추가한 것은 무슨 동작을 할까요?

otter66 replied:
앱이 종료된 후에도 작동시키기 위한 방법을 인터넷을 통해 찾다보니 위와 같은 방법을 찾게 되었습니다. 
(잘못된 정보를 찾은 것 같습니다...!🥲)
android:process=":remote"를 하게되면 프로세스를 분리해 다른 프로세스를 사용하게 된다는 의미인 것 같습니다.

하지만 이 부분에 대해 좀 더 찾아보니 
Manifast에 receiver를 추가하는 것은 정적으로 등록한 것이기 때문에 
앱이 실행중이지 않은 상태에도 프로세스가 뜨고 이벤트를 처리한다는 것을 알았습니다! 

지금의 상태에서는 불필요한 코드인 것 같습니다! 이와 관련된 부분들을 수정하였습니다!

laco-dev commented on app/src/main/java/woowacourse/movie/MyApplication.kt:
```diff
@@ -0,0 +1,14 @@
+
+import android.app.Application
+
+class MyApplication : Application() {
+    companion object {
+        lateinit var prefs: PreferenceUtil
+    }
+
+    override fun onCreate() {
+        prefs = PreferenceUtil(applicationContext)
+        super.onCreate()
+    }
+}
```
Context 의존성을 위해서 Application에서 PreferenceUtil을 주입하도록 잘 작성 해주셨습니다.
하지만, 생각 해 볼 점은 이 인스턴스가 MyApplication이 가지고 있어야 하는가? 에 대해서인데요.
만약, Context가 필요한 다른 모든 싱글턴 객체가 있다면 전부다 MyApplication.xxx로 사용해야 할 것입니다.

아직은 주입 관련된 내용을 학습하기 전이기 때문에, 지금은 아래와 같은 정도로 개선 해 보면 좋겠습니다.

```kotlin
val prefUtil = PreferenceUtil.getInstance() 
```

laco-dev commented on app/src/main/java/woowacourse/movie/PreferenceUtil.kt:
```diff
@@ -0,0 +1,20 @@
+import android.content.SharedPreferences
+
+class PreferenceUtil(context: Context) {
+    private val prefs: SharedPreferences =
+        context.getSharedPreferences(
+            "prefs_name",
+            Context.MODE_PRIVATE + Context.MODE_MULTI_PROCESS
+        )
```
특별한 경우가 아니면, Preferences를 가져올 때는 하나의 저장소로 관리를 합니다.
위 처럼 프리퍼런스를 초기화 하는 이유가 있을까요?

https://developer.android.com/guide/topics/ui/settings/use-saved-values?hl=ko#reading_preference_values

otter66 replied:
이 부분은 xml에서 process를 분리시키도록 한 것 때문에 프리퍼런스를 사용할 수 없어 추가되었습니다.
하나의 저장소를 사용하도록 수정하였습니다!! 감사합니다!

laco-dev commented on app/src/main/java/woowacourse/movie/data/TicketsRepository.kt:
```diff
@@ -0,0 +1,12 @@
+
+    fun addTicket(ticket: TicketsState) = this.tickets.add(ticket)
+
+    private val tickets: MutableList<TicketsState> = mutableListOf()
```
https://kotlinlang.org/docs/coding-conventions.html#class-layout

프로퍼티는 항상 함수보다 위에 위치해야 합니다.

otter66 replied:
변수의 길이가 함수보다 길어져 함수 뒤로 빼두었으나 추후에 기능이 추가될 때 가독성이 문제가 될 수 있겠네요! 
수정하였습니다! 이런 부분에서도 컨벤션을 잊지 않겠습니다!

laco-dev commented on app/src/main/res/values/dimens.xml:
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="utf-8"?>
+<resources>
+    <dimen name="title_small">16sp</dimen>
+
+    <dimen name="margin_small">12dp</dimen>
```
어떤 값들은 텍스트 크기를 리소스로 사용하고, 어떤 값들은 하드코딩을 하는 등의 일관성이 없습니다.

리소스로 만든 목적은 무엇일까요?

laco-dev commented on app/src/main/java/woowacourse/movie/util/Permissions.kt:
```diff
@@ -0,0 +1,24 @@
+import android.content.pm.PackageManager
+import androidx.core.content.ContextCompat
+
+fun Activity.hasPermissions(permissions: Array<String>): Boolean {
```
이 기능은 Activity의 확장함수로 만들어 지는 것이 맞을까요?
프래그먼트에서는 권한을 확인할 수 없나요?

laco-dev commented on app/src/main/java/woowacourse/movie/util/NotificationExt.kt:
```diff
@@ -0,0 +1,73 @@
+        .build()
+}
+
+fun Context.sendNotification(
+    iconResId: Int,
+    contentTitle: String,
+    contentText: String,
+    cancelable: Boolean,
+): Int {
```
비슷하게 Notification에 들어가는 내용이 많아질수록 함수의 파라미터도 복잡해집니다.

Notification을 만들기 위한 필요한 내용들을, 객체로 만들어서 전달 해보세요

laco-dev commented on app/src/main/java/woowacourse/movie/util/NotificationExt.kt:
```diff
@@ -0,0 +1,73 @@
+import androidx.core.app.NotificationManagerCompat
+import java.util.concurrent.atomic.AtomicInteger
+
+const val CHANNEL_ID = "movie_channel"
+const val CHANNEL_NAME = "movie_channel_name"
+const val CHANNEL_DESCRIPTION = "This is a movie channel"
+
+val atomicNotifyId = AtomicInteger(0x001)
```
이 값들이 외부로 노출될 필요가 있을까요?

laco-dev commented on app/src/main/java/woowacourse/movie/util/NotificationExt.kt:
```diff
@@ -0,0 +1,73 @@
+        != PackageManager.PERMISSION_GRANTED
+    ) {
+        Log.d("NotificationExt", "sendNotification: no permission")
+        return -1
```
사용하는 곳에서 -1을 왜 반환하는지 알기 어려울 것 같습니다.

상수로 만들어서 관리를 하거나 명확하가 인지할 수 있도록 해보세요

laco-dev commented on app/src/main/java/woowacourse/movie/ui/main/MainActivity.kt:
```diff
@@ -1,56 +1,101 @@
-            }
-        )
-    }
+    private val rv: FragmentContainerView by lazy { findViewById(R.id.container) }
```
rv가 무엇인가요?

laco-dev commented on app/src/main/java/woowacourse/movie/ui/main/MainActivity.kt:
```diff
@@ -1,56 +1,101 @@
         setContentView(R.layout.activity_main)
-        movieListView.adapter = adapter
+
+        movieListFragment =
+            supportFragmentManager.findFragmentByTag(MOVIE_LIST_TAG) as? MovieListFragment
+                ?: MovieListFragment()
+
+        reservationListFragment =
+            supportFragmentManager.findFragmentByTag(RESERVATION_LIST_TAG) as? ReservationListFragment
+                ?: ReservationListFragment()
+
+        settingFragment =
+            supportFragmentManager.findFragmentByTag(SETTING_TAG) as? SettingFragment
+                ?: SettingFragment()
```
반복되는 코드는 어떻게 한다?

otter66 replied:
줄여야 합니다!!! 🫡

laco-dev commented on app/src/main/java/woowacourse/movie/ui/main/MainActivity.kt:
```diff
@@ -1,56 +1,101 @@
+    private fun initListener() {
+        bottomNavigation.setOnItemSelectedListener {
+            when (it.itemId) {
+                R.id.reservation_list_item -> {
+                    supportFragmentManager.beginTransaction()
+                        .show(reservationListFragment)
+                        .hide(movieListFragment)
+                        .hide(settingFragment)
+                        .commit()
+                }
+                R.id.movie_list_item -> {
+                    supportFragmentManager.beginTransaction()
+                        .hide(reservationListFragment)
+                        .show(movieListFragment)
+                        .hide(settingFragment)
+                        .commit()
+                }
+                R.id.setting_item -> {
+                    supportFragmentManager.beginTransaction()
+                        .hide(reservationListFragment)
+                        .hide(movieListFragment)
+                        .show(settingFragment)
+                        .commit()
+                }
+            }
```
한번에 하나의 프래그먼트만 화면에 보입니다.
그렇다면 add + hide + show 로 관리하는 방법이 맞을까요?
아니면 replace로 관리하는 방법이 맞을까요?

지금의 경우 프래그먼트 갯수가 늘어나면 위 코드도 비약적으로 늘어나야 합니다.

otter66 replied:
.replace(fragmentContainerView.id ...)이 중복되기도 하고, 
단순히 화면이 보이지 않는 것이기 때문에 add + hide + show로 하였습니다. 

하지만 이렇게 바꾸고나니 오히려 중복되는 코드가 많아졌고, 
한 화면에 하나의 프레그먼트가 보이고, 해당 프래그먼트와 같은 위치에서 프래그먼트가 변경되는 이 상황은
말씀해주신 것처럼 replace가 관리하기에 좋을 것 같습니다!

add + hide + show는 한 화면에 fragment가 여럿 보일 때, 
상황에 따라 일부 화면의 fragment가 보이거나 보이지 않아야할 때 적합할 것 같습니다!

laco-dev commented on app/src/main/java/woowacourse/movie/ui/main/MainActivity.kt:
```diff
@@ -1,56 +1,101 @@
+        }
+    }
+
+    private val requestPermissionLauncher = registerForActivityResult(
+        ActivityResultContracts.RequestPermission()
+    ) { isGranted: Boolean ->
```
마찬가지로 프로퍼티는 함수보다 위에 위치해주세요.

그런데 이 코드는 응답 이후 Callback이 아무런 동작을 하지 않네요

laco-dev commented on app/src/main/java/woowacourse/movie/ui/fragment/setting/SettingFragment.kt:
```diff
@@ -0,0 +1,55 @@
+class SettingFragment : Fragment() {
+
+    private var switch: SwitchCompat? = null
+    override fun onCreateView(
+        inflater: LayoutInflater,
+        container: ViewGroup?,
+        savedInstanceState: Bundle?
+    ): View? {
+        // Inflate the layout for this fragment
+        return inflater.inflate(R.layout.fragment_setting, container, false)
+    }
```
프래그먼트의 생성자로 layout 아이디를 넘기는 방법이 있습니다.

laco-dev commented on app/src/main/java/woowacourse/movie/ui/fragment/setting/SettingFragment.kt:
```diff
@@ -0,0 +1,55 @@
+        super.onViewCreated(view, savedInstanceState)
+        switch = view.findViewById(R.id.notification_switch)
+        var switchValue = false
+        context?.let {
```
함수의 indent를 줄이는 것은 레벨 2에서도 모두 동일합니다.

함수 분리, early return등 지금까지 해온 경험들을 활용해서 가독성을 개선 해주세요

laco-dev commented on app/src/main/java/woowacourse/movie/ui/fragment/reservationList/ReservationListFragment.kt:
```diff
@@ -0,0 +1,62 @@
+
+class ReservationListFragment : Fragment() {
+
+    private lateinit var reservationRecyclerView: RecyclerView
```
View를 nullable로 관리하는 곳도 있고 lateinit으로 관리하는 곳도 있네요.
각각 어떤 기준으로 나누어 진 것일까요?

laco-dev commented on app/src/main/java/woowacourse/movie/ui/fragment/movieList/MovieListFragment.kt:
```diff
@@ -0,0 +1,70 @@
+    }
+
+    private fun navigateMovieDetail(movie: MovieState) {
+        val intent = Intent(activity, MovieDetailActivity::class.java)
+        intent.putExtra(MainActivity.KEY_MOVIE, movie)
+        startActivity(intent)
```
https://github.com/PRNDcompany/android-style-guide/blob/main/Java.md#activity

액티비티의 Intent를 가져오는 것도 객체지향 관점에서 보면 스스로 해야할 역할일 수 있습니다.
위 내용은 일반적으로 사용되는 패턴으로 참고해서 개선 해보세요

otter66 replied:
여러곳에서 하나의 액티비티를 띄우기 위해 같은 코드가 생길 것 같다는 고민이 있었습니다! 
코드의 중복을 줄일 수 있고, intent를 하는 과정을 간략하게 표현할 수 있을 것 같습니다!
감사합니다!!!!

otter66 replied:
intent할 때 전달해야 할 date들도 명시할 수 있겠네요! 여러모로 좋은 방법인 것 같네요!! 🤔
라코데브님 최고!

laco-dev commented on app/src/main/java/woowacourse/movie/ui/confirm/ReservationConfirmActivity.kt:
```diff
@@ -48,4 +56,33 @@ class ReservationConfirmActivity : BackKeyActionBarActivity() {
+                tickets.dateTime.minute
+            )
+        }
+        val alarmManager: AlarmManager = getSystemService(ALARM_SERVICE) as AlarmManager
+        val alarmIntent =
+            Intent(this, AlarmReceiver::class.java).apply { putExtra("a", tickets) }.let { intent ->
+                PendingIntent.getBroadcast(
+                    this,
+                    tickets.hashCode(),
+                    intent,
+                    PendingIntent.FLAG_UPDATE_CURRENT + PendingIntent.FLAG_IMMUTABLE
+                )
+            }
+
+        alarmManager.setExactAndAllowWhileIdle(
+            AlarmManager.RTC_WAKEUP,
+            calendar.timeInMillis - 30 * 60 * 1000,
+            alarmIntent
+        )
```
알람과 알림등에 대한 코드가 많이 복잡하게 느껴집니다.

별도의 객체로 만들어서 관리 해보면 어떨까요?

laco-dev reviewed (changes requested):
1, 2 단계 미션 잘 해주셨습니다.

도메인 로직과 달리 안드로이드에 더 친해지는 과정으로 조금씩 발전해 나가는 모습이 좋았습니다.
다만 우선 동작하는데 집중하다보니 많은 부분에서 반복되는 코드가 생겨났습니다.

새롭게 레벨 2를 구현하는 것이 아닌 레벨 1에서 배워온 경험들을 잘 살려주시는게 중요합니다.

Commit:
style: 컨벤션과 다른 메서드와 프로퍼티 순서 수정

Commit:
refactor: 일관성 없는 상수화 제거

Commit:
refactor: 패키지 구조 변경

- ui 안의 main에 있던 파일들의 구조를 변경
- ui/main 내부의 패키지를 외부로 옮김

Commit:
refactor: 불필요하게 노출되는 변수들의 접근제어자 수정

Commit:
refactor: 불필요 로그 제거

Commit:
refactor: 메서드 및 변수명 변경

Commit:
refactor: 불필요 함수 제거

Commit:
refactor(mainActivity): fragment 관리 방식 변경

- add, hide, show 방식에서 replace하는 방식으로 수정

Commit:
refactor(mainActivity): 불필요 코드 제거 및 컨벤션 적용

Commit:
refactor(fragment): fragment view 생성 방식 변경

Commit:
refactor(notification): 리시버의 프로세스를 분리하지 않도록 수정

- 기존에 프로세스를 분리한 것은 앱이 실행중이지 않은 상황에도 리시버를 작동시키기 위함이었음
- 리시버를 xml로 등록해 불필요한 작업이기에 제거.

Commit:
fix: 예약 결과 화면 회전시 여러번 저장되는 문제 해결

- 예약 결과 화면에서 저장을 진행하며 발생하는 문제
- 예약 확인 버튼을 눌렀을 때 예약 관련 이벤트(데이터 추가, 알림 설정)가 진행되도록 수정

Commit:
refactor: 프래그먼트에서 일관성 없는 view nullable 변경

- 모든 프래그먼트에서 view가 null을 가지지 않도록 수정
- 그 외에도 receiver의 패키지, intent key 상수화 등 수정

Commit:
refactor(notification): 알림 설정 값들을 객체로 관리하도록 수정

Commit:
refactor(notification): 알림 반환값 상수화

Commit:
refactor(fragment): 프래그먼트 선언 및 관리 방식 변경

Commit:
style(setting): 코드 가독성 개선

Commit:
refactor(notification): 패키지 분리

Commit:
refactor(activity): intent, startActivity 분리

Commit:
fix(notification): 알람 안울리는 문제 해결

- intent key 값이 다른 문제
- 테스트를 위한 tickets value 제거

otter66 commented:
피드백을 전부 반영하고 싶었으나, 학습 시간이 너무 길어지는 것 같아 우선 리뷰 요청드립니다!
스스로 좀 더 찾아본 후 반영하지 못한 부분들을 추가하도록 하겠습니다...!

laco-dev commented on app/src/main/java/woowacourse/movie/main/MainActivity.kt:
```diff
@@ -0,0 +1,95 @@
+    private fun bottomNavigationItemClickEvent(menuItem: MenuItem) {
+        when (menuItem.itemId) {
+            R.id.reservation_list_item -> {
+                supportFragmentManager.commit {
+                    replace(fragmentContainerView.id, reservationListFragment)
+                }
```
이 코드를 하나의 함수로 분리해 볼 수도 있지 않을까요?

laco-dev commented on app/src/main/java/woowacourse/movie/ui/fragment/setting/SettingFragment.kt:
```diff
@@ -1,55 +1,42 @@
+            switchCompat.isChecked = false
+            Toaster.showToast(requireContext(), "알림 권한을 허용해주세요.")
+        }
+        sharedPreference.setBoolean(NOTIFICATIONS, switchCompat.isChecked)
```
이 키값에 대한 변경을 여러군데에서 한다면 어떻게 될까요?
모든 화면이 `SettingFragment.NOTIFICATIONS`를 사용해야 할까요?

동일한 키값에 대해서는 PreferenceUtil에서 관리해보도록 해보세요.

laco-dev commented on app/src/main/java/woowacourse/movie/ui/seat/SeatSelectActivity.kt:
```diff
@@ -71,16 +81,18 @@ class SeatSelectActivity : BackKeyActionBarActivity() {
+
+    private fun reserveMovie(seats: List<SeatPositionState>) {
+        val tickets: TicketsState = TicketsState.from(ticketOptState, seats)
+        TicketsRepository.addTicket(tickets)
+        navigateReservationConfirmActivity(tickets)
+        setNotification(tickets)
```
아래 내용은 지금 고려하지는 않아도 됩니다. 🙏🏻 

티켓을 예약한다는 비즈니스 로직은 항상 성공하지는 않을 수 있어요.
그렇기 때문에 비즈니스 로직을 호출하는 부분 (Repository.addXXX) 과, UI 로직 (navigate)이 문제없이 잘 호출되는지 고민하는 것도 중요합니다.

laco-dev commented on app/src/main/java/woowacourse/movie/ui/seat/SeatSelectActivity.kt:
```diff
@@ -99,7 +111,43 @@ class SeatSelectActivity : BackKeyActionBarActivity() {
+                )
+            }
+
+        alarmManager.setExactAndAllowWhileIdle(
+            AlarmManager.RTC_WAKEUP,
+            calendar.timeInMillis - 30 * 60 * 1000,
+            alarmIntent
+        )
```
알람에 대한 실제 구현도, 별도 객체로 분리 해보면 어떨까요?

현재는 사용하는 입장에서 AlarmManager라는 기능에 대해 자세히 알아야 합니다.

laco-dev commented on app/src/main/java/woowacourse/movie/util/PreferenceUtil.kt:
```diff
@@ -1,14 +1,11 @@
-            "prefs_name",
-            Context.MODE_PRIVATE + Context.MODE_MULTI_PROCESS
-        )
+        context.getSharedPreferences("prefs_name", Context.MODE_PRIVATE)
```
프리퍼런스와 같은 데이터 접근 객체는 싱글톤으로 관리가 되고는 하는데요.
현재 구현에서는 위 코드에 접근하기 위해서 매번 PreferenceUtil 인스턴스를 생성해야 합니다.

기회가 되신다면, 싱글톤 관련 개념을 학습 해보시고 클래스를 싱글톤으로 가져오는 방법을 공부 해보시면 좋겠습니다.

object가 아닙니다!

laco-dev commented on app/src/main/java/woowacourse/movie/util/notification/NotificationSettings.kt:
```diff
@@ -0,0 +1,11 @@
+package woowacourse.movie.util.notification
```
notification은.. 과연 유틸일까요?

laco-dev reviewed (approved):
피드백 반영 고생 많으셨습니다.

다음 아키텍처 미션에 집중하기 위해서, 이번 단계는 마무리 하도록 하겠습니다.
남겨드린 의견은 고민 해 보시고 다음 단계에 반영 해 주시면 됩니다.
